language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/CodeActionRouter.java

```diff
@@ -99,25 +99,29 @@
                         .isWithinRange(ctx.cursorPosition(), CommonUtil.toRange(diag.location().lineRange()))
                 )
                 .forEach(diagnostic -> {
-                    DiagBasedPositionDetails positionDetails = computePositionDetails(syntaxTree, diagnostic, ctx);
-                    codeActionProvidersHolder.getActiveDiagnosticsBasedProviders(ctx).forEach(provider -> {
-                        try {
-                            // Check whether the code action request has been cancelled
-                            // in order to avoid unnecessary calculations
-                            ctx.checkCancelled();
-                            
-                            List<CodeAction> codeActionsOut = provider.getDiagBasedCodeActions(diagnostic,
-                                                                                               positionDetails, ctx);
-                            if (codeActionsOut != null) {
-                                codeActionsOut.forEach(codeAction ->
-                                        TelemetryUtil.addReportFeatureUsageCommandToCodeAction(codeAction, provider));
-                                codeActions.addAll(codeActionsOut);
+                        DiagBasedPositionDetails positionDetails = computePositionDetails(syntaxTree, diagnostic, ctx);
+                        codeActionProvidersHolder.getActiveDiagnosticsBasedProviders(ctx)
+                                .stream().filter(provider -> provider.validate(diagnostic, positionDetails, ctx))
```

Comment:
What if we do this in a single run? Without filtering first and then iterating, what if we do the validation check before executing, in the sense within the iteration